shell/platform/embedder/embedder.h

```diff
@@ -401,6 +404,57 @@ typedef struct {
   VoidCallback destruction_callback;
 } FlutterOpenGLFramebuffer;
 
+typedef bool (*FlutterOpenGLSurfaceCallback)(void* /* user data */,
+                                             bool* /* opengl state changed */);
+
+typedef struct {
+  /// The size of this struct. Must be sizeof(FlutterOpenGLSurface).
+  size_t struct_size;
+
+  /// User data to be passed to the make_current, clear_current and
+  /// destruction callback.
+  void* user_data;
+
+  /// Callback invoked (on an engine managed thread) that asks the embedder to
+  /// make the surface current.
+  ///
+  /// Should return true if the operation succeeded, false if the surface could
+  /// not be made current and rendering should be cancelled.
+  ///
+  /// The second parameter 'opengl state changed' should be set to true if
+  /// any OpenGL API state has changed and flutter should not assume any native
+  /// API state is the same as before this callback was called.
+  ///
+  /// In that case, flutter will invalidate the internal OpenGL API state cache,
+  /// which is a somewhat expensive operation.
+  ///
+  /// @attention required. (non-null)
+  FlutterOpenGLSurfaceCallback make_current_callback;
+
+  /// Callback invoked (on an engine managed thread) when the current surface
+  /// can be cleared.
+  ///
+  /// Should return true if the operation succeeded, false if an error ocurred.
+  /// That error will be logged but otherwise not handled by the engine.
+  ///
+  /// The second parameter 'opengl state changed' is the same as with the
+  /// @ref make_current_callback.
+  ///
+  /// The embedder might clear the surface here after it was previously made
+  /// current. That's not required however, it's also possible to clear it in
+  /// the destruction callback. There's no way to signal opengl state
+  /// changes in the destruction callback though.
+  ///
+  /// @attention required. (non-null)
+  FlutterOpenGLSurfaceCallback clear_current_callback;
+
+  /// Callback invoked (on an engine managed thread) that asks the embedder to
+  /// collect the surface.
```

Comment:
nit: This is fine, but being super pedantic, I suspect something like "that implements embedder-specific surface collection." would be a tiny bit clearer since the embedder itself will wire this up.

Reply:
there's a lot of places that use this wording though: https://github.com/flutter/engine/blob/eafe147b3875ef6cb1290525b9b803d89ef11df9/shell/platform/embedder/embedder.h#L377-L379

so if people see that wording they might recognize it and see "oh it's just that same callback but for a surface"

but yeah, maybe it could be improved, what's also left a bit unclear is the timing (which of the overall backing store and opengl texture/fbo/surface destruction callback is called first), if you can leave them empty and specify a NULL value why you would want to have two callbacks in general.

Reply:
Hmm if we're using that phrasing elsewhere then leave as-is for consistency and we can fix it across the board later.